File: Parser.py
import antlr4
from Error import printCompilerMessage, ErrorType
from Location import Location
from grammar import HazeLexer, HazeParser
from antlr4.error.ErrorListener import ErrorListener
from antlr4 import InputStream


class HazeErrorListener(ErrorListener):
    def __init__(self, filename: str):
        self.filename = filename

    def syntaxError(self, recognizer, offendingSymbol, line, column, msg, e):
        printCompilerMessage(
            Location(self.filename, line, column), ErrorType.Error, "SyntaxError", msg
        )


class Parser:
    def __init__(self, filename: str):
        with open(filename, "r") as fs:
            self.inputStream = InputStream(fs.read())
        self.lexer = HazeLexer.HazeLexer(self.inputStream)
        self.errorListener = HazeErrorListener(filename)
        self.lexer.removeErrorListeners()
        self.lexer.addErrorListener(self.errorListener)
        self.tokens = antlr4.CommonTokenStream(self.lexer)
        self.parser = HazeParser.HazeParser(self.tokens)
        self.parser.removeErrorListeners()
        self.parser.addErrorListener(self.errorListener)

    def parse(self):
        if self.parser.getNumberOfSyntaxErrors() != 0:
            return None
        self.ast = self.parser.prog()
        if self.parser.getNumberOfSyntaxErrors() != 0:
            return None
        return self.ast

File: src/parser.ts
import { HazeLexer } from "./grammar/autogen/HazeLexer";
import { HazeParser } from "./grammar/autogen/HazeParser";
import { SourceLoc, printErrorMessage } from "./Errors";
import { BaseErrorListener, CharStream, CommonTokenStream } from "antlr4ng";

class HazeErrorListener extends BaseErrorListener {
  filename: string;

  constructor(filename: string) {
    super();
    this.filename = filename;
  }

  syntaxError(
    recognizer: any,
    offendingSymbol: any,
    line: number,
    column: number,
    msg: string,
    e: any,
  ) {
    printErrorMessage(
      msg,
      new SourceLoc(this.filename, line, column),
      "SyntaxError",
    );
  }
}

export class Parser {
  parser?: HazeParser;

  constructor() {}

  async parseFile(filename: string) {
    const file = Bun.file(filename);
    const text = await file.text();
    return await this.parse(text, filename);
  }

  async parse(text: string, errorListenerFilename: string) {
    const errorListener = new HazeErrorListener(errorListenerFilename);
    let inputStream = CharStream.fromString(text);
    let lexer = new HazeLexer(inputStream);
    lexer.removeErrorListeners();
    lexer.addErrorListener(errorListener);

    let tokenStream = new CommonTokenStream(lexer);
    this.parser = new HazeParser(tokenStream);
    this.parser.removeErrorListeners();
    this.parser.addErrorListener(errorListener);

    if (this.parser.numberOfSyntaxErrors != 0) {
      return;
    }
    const ast = this.parser.prog();
    if (this.parser.numberOfSyntaxErrors != 0) {
      return;
    }
    return ast;
  }
}
